### PR TITLE
PYI-589

### PR DESCRIPTION
**Pull request by SunilDGDS**
### What changed
Verify signature of the shared attributes JWT 
Added happy & unhappy paths tests

- [PYI-589](https://govukverify.atlassian.net/browse/PYI-589)

### CODE REVIEW

File: lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java [hunk 1]
```diff
@@ -13,9 +15,14 @@
 import com.nimbusds.oauth2.sdk.util.StringUtils;
 import uk.gov.di.ipv.cri.passport.library.service.ConfigurationService;
 
+import java.io.ByteArrayInputStream;
```
Comment (dannyhayes-gds) [inline lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java:18]: Could you remove some of these unused imports here

File: lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java [hunk 2]
```diff
@@ -53,22 +60,33 @@ public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent in
...
 
         try {
+            SignedJWT signedJWT = SignedJWT.parse(input.getBody());
             Certificate clientCert = configurationService.getClientCert(clientId);
 
-        } catch (CertificateException e) {
-            LOGGER.log(Level.WARNING, "Failed to retrieve certificate");
-            return ApiGatewayResponseGenerator.proxyJsonResponse(BAD_REQUEST, ErrorResponse.FAILED_TO_RETRIEVE_CERTIFICATE);
-        }
+            if (isInvalidSignature(SignedJWT.parse(input.getBody()), clientCert)) {
```
Comment (dannyhayes-gds) [inline lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java:66]: You don't need to parse the JWT again here, instead just pass in the signedJwt from above

```suggestion
            if (isInvalidSignature(signedJwt, clientCert)) {
```

File: lambdas/jwtverification/src/test/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandlerTest.java [hunk 3]
```diff
@@ -10,6 +10,9 @@
 import com.nimbusds.jose.JWSHeader;
 import com.nimbusds.jose.JWSSigner;
 import com.nimbusds.jose.crypto.MACSigner;
```
Comment (dannyhayes-gds) [inline lambdas/jwtverification/src/test/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandlerTest.java:12]: There are a few unused imports here as well

File: lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java (hunk 1 above)
Comment (SunilDGDS) [reply lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java]: Optimised

File: lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java (hunk 2 above)
Comment (SunilDGDS) [reply lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java]: Refactored

File: lambdas/jwtverification/src/test/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandlerTest.java (hunk 3 above)
Comment (SunilDGDS) [reply lambdas/jwtverification/src/test/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandlerTest.java]: Optimised

File: lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java [hunk 4]
```diff
@@ -53,22 +60,33 @@ public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent in
...
+            LOGGER.log(Level.WARNING, "Failed to parse the shared attributes JWT", e);
             return ApiGatewayResponseGenerator.proxyJsonResponse(BAD_REQUEST, ErrorResponse.FAILED_TO_PARSE_SHARED_ATTRIBUTES_JWT);
+        } catch (CertificateException | JOSEException e) {
+            LOGGER.log(Level.WARNING, "Failed to verify the signature of the JWT", e);
+            return ApiGatewayResponseGenerator.proxyJsonResponse(BAD_REQUEST, ErrorResponse.FAILED_TO_VERIFY_SIGNATURE);
         }
     }
+
+    private boolean isInvalidSignature(SignedJWT signedJWT, Certificate clientCertificate)
+            throws CertificateException, JOSEException {
```
[the remark this answers is not in the document]
Comment (SunilDGDS) [reply lambdas/jwtverification/src/main/java/uk/gov/di/ipv/cri/passport/jwtverification/JwtVerificationHandler.java]: Refactored

**Review by dannyhayes-gds — APPROVED** [review]

**Pull request merged**